diff --git a/QDMA/linux-kernel/driver/src/cdev.c b/QDMA/linux-kernel/driver/src/cdev.c
index bd7d681..76a9204 100755
--- a/QDMA/linux-kernel/driver/src/cdev.c
+++ b/QDMA/linux-kernel/driver/src/cdev.c
@@ -2,7 +2,7 @@
  * This file is part of the Xilinx DMA IP Core driver for Linux
  *
  * Copyright (c) 2017-2022, Xilinx, Inc. All rights reserved.
- * Copyright (c) 2022-2023, Advanced Micro Devices, Inc. All rights reserved.
+ * Copyright (c) 2022-2024, Advanced Micro Devices, Inc. All rights reserved.
  *
  * This source code is free software; you can redistribute it and/or modify it
  * under the terms and conditions of the GNU General Public License,
@@ -51,6 +51,9 @@ struct xlnx_phy_dev {
 	struct list_head list_head;	/**< board list */
 	unsigned int major;	/**< major number per board */
 	unsigned int device_bus;	/**< PCIe device bus number per board */
+#ifdef CONFIG_PCI_DOMAINS_GENERIC
+	unsigned int device_bus_domain; /**< PCIe bus domain per board */
+#endif
 	unsigned int dma_device_index;
 };
 
@@ -567,12 +570,40 @@ static ssize_t cdev_aio_read(struct kiocb *iocb, const struct iovec *io,
 #if KERNEL_VERSION(3, 16, 0) <= LINUX_VERSION_CODE
 static ssize_t cdev_write_iter(struct kiocb *iocb, struct iov_iter *io)
 {
-	return cdev_aio_write(iocb, io->iov, io->nr_segs, iocb->ki_pos);
+#ifdef RHEL_RELEASE_VERSION
+	#if RHEL_RELEASE_VERSION(9, 4) > RHEL_RELEASE_CODE
+		return cdev_aio_write(iocb, io->iov, io->nr_segs, iocb->ki_pos);
+	#else
+		return cdev_aio_write(iocb, iter_iov(io), io->nr_segs,
+						         iocb->ki_pos);
+	#endif
+#else
+	#if KERNEL_VERSION(6, 4, 0) > LINUX_VERSION_CODE
+		return cdev_aio_write(iocb, io->iov, io->nr_segs, iocb->ki_pos);
+	#else
+		return cdev_aio_write(iocb, iter_iov(io), io->nr_segs,
+							 iocb->ki_pos);
+	#endif
+#endif
 }
 
 static ssize_t cdev_read_iter(struct kiocb *iocb, struct iov_iter *io)
 {
-	return cdev_aio_read(iocb, io->iov, io->nr_segs, iocb->ki_pos);
+#ifdef RHEL_RELEASE_VERSION
+	#if RHEL_RELEASE_VERSION(9, 4) > RHEL_RELEASE_CODE
+		return cdev_aio_read(iocb, io->iov, io->nr_segs, iocb->ki_pos);
+	#else
+		return cdev_aio_read(iocb, iter_iov(io), io->nr_segs,
+							iocb->ki_pos);
+	#endif
+#else
+	#if KERNEL_VERSION(6, 4, 0) > LINUX_VERSION_CODE
+		return cdev_aio_read(iocb, io->iov, io->nr_segs, iocb->ki_pos);
+	#else
+		return cdev_aio_read(iocb, iter_iov(io), io->nr_segs,
+							iocb->ki_pos);
+	#endif
+#endif
 }
 #endif
 
@@ -626,8 +657,7 @@ int qdma_cdev_create(struct qdma_cdev_cb *xcb, struct pci_dev *pdev,
 	int rv;
 	unsigned long *priv_data;
 
-	xcdev = kzalloc(sizeof(struct qdma_cdev) + strlen(qconf->name) + 1,
-			GFP_KERNEL);
+	xcdev = kzalloc(sizeof(struct qdma_cdev), GFP_KERNEL);
 	if (!xcdev) {
 		pr_err("%s failed to allocate cdev %lu.\n",
 		       qconf->name, sizeof(struct qdma_cdev));
@@ -647,7 +677,7 @@ int qdma_cdev_create(struct qdma_cdev_cb *xcb, struct pci_dev *pdev,
 			&xcdev->c2h_qhndl : &xcdev->h2c_qhndl;
 	*priv_data = qhndl;
 	xcdev->dir_init = (1 << qconf->q_type);
-	strcpy(xcdev->name, qconf->name);
+	strscpy(xcdev->name, qconf->name, sizeof(xcdev->name));
 
 	xcdev->minor = minor;
 	if (xcdev->minor >= xcb->cdev_minor_cnt) {
@@ -743,6 +773,10 @@ int qdma_cdev_device_init(struct qdma_cdev_cb *xcb)
 	xdev = (struct xlnx_dma_dev *)xcb->xpdev->dev_hndl;
 	list_for_each_entry_safe(phy_dev, tmp, &xlnx_phy_dev_list, list_head) {
 		if (phy_dev->device_bus == xcb->xpdev->pdev->bus->number &&
+#ifdef CONFIG_PCI_DOMAINS_GENERIC
+			phy_dev->device_bus_domain ==
+				xcb->xpdev->pdev->bus->domain_nr &&
+#endif
 			phy_dev->dma_device_index == xdev->dma_device_index) {
 			xcb->cdev_major = phy_dev->major;
 			mutex_unlock(&xlnx_phy_dev_mutex);
@@ -768,6 +802,9 @@ int qdma_cdev_device_init(struct qdma_cdev_cb *xcb)
 
 	new_phy_dev->major = xcb->cdev_major;
 	new_phy_dev->device_bus = xcb->xpdev->pdev->bus->number;
+#ifdef CONFIG_PCI_DOMAINS_GENERIC
+	new_phy_dev->device_bus_domain = xcb->xpdev->pdev->bus->domain_nr;
+#endif
 	new_phy_dev->dma_device_index = xdev->dma_device_index;
 	xlnx_phy_dev_list_add(new_phy_dev);
 
@@ -780,7 +817,19 @@ int qdma_cdev_device_init(struct qdma_cdev_cb *xcb)
 
 int qdma_cdev_init(void)
 {
-	qdma_class = class_create(THIS_MODULE, QDMA_CDEV_CLASS_NAME);
+#ifdef RHEL_RELEASE_VERSION
+	#if RHEL_RELEASE_VERSION(9, 4) > RHEL_RELEASE_CODE
+		qdma_class = class_create(THIS_MODULE, QDMA_CDEV_CLASS_NAME);
+	#else
+		qdma_class = class_create(QDMA_CDEV_CLASS_NAME);
+	#endif
+#else
+	#if KERNEL_VERSION(6, 4, 0) > LINUX_VERSION_CODE
+		qdma_class = class_create(THIS_MODULE, QDMA_CDEV_CLASS_NAME);
+	#else
+		qdma_class = class_create(QDMA_CDEV_CLASS_NAME);
+	#endif
+#endif
 	if (IS_ERR(qdma_class)) {
 		pr_err("%s: failed to create class 0x%lx.",
 			QDMA_CDEV_CLASS_NAME, (unsigned long)qdma_class);
diff --git a/QDMA/linux-kernel/driver/src/cdev.h b/QDMA/linux-kernel/driver/src/cdev.h
index 371e528..5512e86 100755
--- a/QDMA/linux-kernel/driver/src/cdev.h
+++ b/QDMA/linux-kernel/driver/src/cdev.h
@@ -90,7 +90,7 @@ struct qdma_cdev {
 	ssize_t (*fp_aiorw)(unsigned long dev_hndl, unsigned long qhndl,
 			unsigned long count, struct qdma_request **reqv);
 	/** name of the character device*/
-	char name[0];
+	char name[QDMA_QUEUE_NAME_MAXLEN];
 };
 
 /**
